lib/ecto/changeset.ex

```diff
@@ -225,33 +225,86 @@ defmodule Ecto.Changeset do
                optional: optional}
   end
 
-  defp process_param(key, kind, params, types, model, {changes, errors}) do
+  defp process_param({key, fun}, kind, params, types, model, acc) do
     {key, param_key} = cast_key(key)
+    type = type!(types, key, fun)
```

Comment:
In other words, if you have a function, assert it is an embed, merge the given function into the embed struct, and then call the code as usual.